# Using Bank Marketing dataset we need to predict whether the customer will subscribe to our servises. Subscription is our target





In [1]:
#Importing the libraries
import pandas as pd

In [2]:
bank_marketing = pd.read_excel('F:\Skillenable\Data frames\Bank marketing.xlsx')
bank_marketing.head()

Age            Job Marital Status  Account Balance  Own House  \
0   59     unemployed        married                0          0   
1   36         Others        married             3057          0   
2   47    blue-collar       divorced              126          1   
3   43     management       divorced              388          1   
4   34  self-employed         single              462          0   

   Personal Loan  No of campaigns  Subscription  
0              0                0             0  
1              0                0             1  
2              0                0             1  
3              0                0             1  
4              0                0             1

In [3]:
bank_marketing.info()
bank_marketing.isnull().sum()   

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4521 entries, 0 to 4520
Data columns (total 8 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Age              4521 non-null   int64 
 1   Job              4521 non-null   object
 2   Marital Status   4521 non-null   object
 3   Account Balance  4521 non-null   int64 
 4   Own House        4521 non-null   int64 
 5   Personal Loan    4521 non-null   int64 
 6   No of campaigns  4521 non-null   int64 
 7   Subscription     4521 non-null   int64 
dtypes: int64(6), object(2)
memory usage: 282.7+ KB


Age                0
Job                0
Marital Status     0
Account Balance    0
Own House          0
Personal Loan      0
No of campaigns    0
Subscription       0
dtype: int64

In [4]:
# crete dummy data for catagorical values

bank_marketing.shape

(4521, 8)

In [3]:
bank_marketing.shape[0]  

4521

In [5]:
cat_col = ['Job', 'Marital Status']
cat_col

['Job', 'Marital Status']

In [6]:
bank_dummy = pd.get_dummies(bank_marketing, columns = cat_col, drop_first= True)
bank_dummy

Age  Account Balance  Own House  Personal Loan  No of campaigns  \
0      59                0          0              0                0   
1      36             3057          0              0                0   
2      47              126          1              0                0   
3      43              388          1              0                0   
4      34              462          0              0                0   
...   ...              ...        ...            ...              ...   
4516   58                0          1              0                0   
4517   38                0          1              0                0   
4518   53             2627          1              1                0   
4519   49              371          1              0                0   
4520   35              280          1              1                0   

      Subscription  Job_blue-collar  Job_management  Job_self-employed  \
0                0                0               0                  0   
1                1                0               0                  0   
2                1                1               0                  0   
3                1                0               1                  0   
4                1                0               0                  1   
...            ...              ...             ...                ...   
4516             0                1               0                  0   
4517             0                0               0                  0   
4518             0                0               0                  1   
4519             0                1               0                  0   
4520             0                1               0                  0   

      Job_unemployed  Marital Status_married  Marital Status_single  
0                  1                       1                      0  
1                  0                       1                      0  
2                  0                       0                      0  
3                  0                       0                      0  
4                  0                       0                      1  
...              ...                     ...                    ...  
4516               0                       1                      0  
4517               0                       1                      0  
4518               0                       1                      0  
4519               0                       1                      0  
4520               0                       1                      0  

[4521 rows x 12 columns]

In [7]:
# create input nad output 
Y = bank_dummy[['Subscription']]
X = bank_dummy.drop(columns=['Subscription'])

In [8]:
# spliting data into test and train

from sklearn.model_selection import train_test_split

X_train,X_test,Y_train,Y_test = train_test_split(X,Y,test_size=0.2,random_state=47)
len(X_train),len(X_test),len(Y_train),len(Y_test)

(3616, 905, 3616, 905)

## 1) What is the accuracy when you use as Random forest?

In [9]:
# creating random forest model
from sklearn.ensemble import RandomForestClassifier

model = RandomForestClassifier(n_estimators=500,random_state=42)
rf = model.fit(X_train,Y_train)


C:\Users\arun\AppData\Local\Temp/ipykernel_16608/2412031796.py:5: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  rf = model.fit(X_train,Y_train)


In [10]:
# Predicting the model on test data
Y_test

Subscription
1034             0
3673             0
1004             0
3489             0
682              0
...            ...
1326             0
2012             0
2519             0
755              0
2917             0

[905 rows x 1 columns]

In [11]:
Y_test['Prediction'] = model.predict(X_test)
Y_test

C:\Users\arun\AppData\Local\Temp/ipykernel_16608/583016460.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  Y_test['Prediction'] = model.predict(X_test)


Subscription  Prediction
1034             0           0
3673             0           0
1004             0           0
3489             0           0
682              0           0
...            ...         ...
1326             0           0
2012             0           0
2519             0           0
755              0           0
2917             0           0

[905 rows x 2 columns]

In [12]:
from sklearn.metrics import confusion_matrix, accuracy_score

print(confusion_matrix(Y_test['Subscription'],Y_test['Prediction']))

[[759  30]
 [106  10]]


In [13]:
print(accuracy_score(Y_test['Subscription'],Y_test['Prediction']))


0.8497237569060774


### the accuracy of Forest tree is 84.97%

## 3) Find the top 4 important factors by using Random Forest. What is your inference?

In [14]:
model.feature_importances_

array([0.29715908, 0.47789397, 0.02710084, 0.01590041, 0.09106062,
       0.01774676, 0.02177484, 0.00957233, 0.00738627, 0.01956348,
       0.0148414 ])

In [15]:
forest_importance = pd.Series(model.feature_importances_,index=X.columns)
forest_importance.sort_values(ascending=False).head(10)

Account Balance           0.477894
Age                       0.297159
No of campaigns           0.091061
Own House                 0.027101
Job_management            0.021775
Marital Status_married    0.019563
Job_blue-collar           0.017747
Personal Loan             0.015900
Marital Status_single     0.014841
Job_self-employed         0.009572
dtype: float64

## 2) What is the accuracy when you use an SVM? How does changing the model affect the accuracy?


In [16]:
# building SVM 
from sklearn.svm import SVC
model1 = SVC(random_state=47)

In [17]:
svc_model = model1.fit(X_train,Y_train)  # fit train data


C:\Users\arun\anaconda3\lib\site-packages\sklearn\utils\validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(*args, **kwargs)


In [19]:
# prediction
Y_test['Prediction'] = svc_model.predict(X_test)
Y_test


C:\Users\arun\AppData\Local\Temp/ipykernel_16608/739399881.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  Y_test['Prediction'] = svc_model.predict(X_test)


Subscription  Prediction
1034             0           0
3673             0           0
1004             0           0
3489             0           0
682              0           0
...            ...         ...
1326             0           0
2012             0           0
2519             0           0
755              0           0
2917             0           0

[905 rows x 2 columns]

In [22]:
# confusion metrix
from sklearn.metrics import confusion_matrix,accuracy_score
print(confusion_matrix(Y_test['Subscription'],Y_test['Prediction']))

[[789   0]
 [116   0]]


In [23]:
# accuracy
print(accuracy_score(Y_test['Subscription'],Y_test['Prediction']))


0.8718232044198895


###  Accuracy of SVM is 87.18 %; accuracy of SVM is higher than Random Forest 

## 4) Which evaluation metrics is important and why?

* Evaluation metrics explain the performance of a model. An important aspect of evaluation metrics is their capability to discriminate among model results.
* following are some evalution metrics
Confusion Matrix
F1 Score
Gain and Lift Charts
Kolmogorov Smirnov Chart
AUC – ROC
Log Loss
Gini Coefficient
Concordant – Discordant Ratio
Root Mean Squared Error
Cross Validation (Not a metric though!)

If we have a tight margin of error then we should follow precision. 